Public partition selection

lovroprepolec opened the pull request:
## Description
Implementing [Issue #4](https://github.com/OpenMined/PipelineDP/issues/4) Public partition selection.   
Remark: [TestPipeline](https://beam.apache.org/documentation/pipelines/test-your-pipeline/#testpipeline) doesn't work for me on the version 2.30.0 (it seems that it doesn't exist). I have used normal Pipeline in tests instead. Let me know if this is an issue.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
Unit tests and the examples.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+        def is_public(col):
+          return col[0] in public_partitions
+
+        VALUES = 0
```
Please add comment before, something like: # define constants for using as keys in CoGroupByKey

Nit: VALUES, IS_PUBLIC = 0, 1     (it looks like here it's simple enough to use multiple assignment)

lovroprepolec replied:
Thanks. Done.

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+        col = col | beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        pp_type = type(public_partitions)
+        if pp_type is list or pp_type is set:
```
Nit: if isinstance(public_partitions, (list, set))

https://docs.python.org/3/library/functions.html#isinstance

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+
+        pp_type = type(public_partitions)
+        if pp_type is list or pp_type is set:
+          return col | beam.Filter(is_public)
```
Please create a set from public_partitions elements, if public_partitions is not set (look-up in a large list  might be too expensive)

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/dp_engine.py:
```diff
@@ -48,7 +48,17 @@ def aggregate(self, col, params: AggregateParams,
-        return col
+        result = col
+
+        # IF no public partitions were specified, return aggregation results
```
IF -> If

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+        if public_partitions is None:
+          raise TypeError("Must provide a valid public_partitions")
+
+        col = col | beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
```
Please add stage description for each PTransform applications (i.e. col | "some description" >> beam.Map() )

Explanation: If description is empty, Beam sets default name, something like "Map". And if there >1 non-named beam.Map it leads to name collision.

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+            # TODO the Issue #4 says this is blocked on other tasks. Revisit
+            # this once unblocked
+            if not values:
+              values = [None]
```
Please replace this line with "return" (returning [None] would break further stages which are in development, later this will be fixed.

lovroprepolec replied:
Done.

lovroprepolec commented on examples/movie_view_ratings.py:
```diff
@@ -95,11 +96,12 @@ def process(self, line):
 
 def compute_on_beam():
     runner = fn_api_runner.FnApiRunner()  # local runner
+    public_partitions=[int(partition) for partition in FLAGS.public_partitions] if FLAGS.public_partitions is not None else None
```
Done.

lovroprepolec commented on examples/movie_view_ratings.py:
```diff
@@ -95,11 +96,12 @@ def process(self, line):
 
 def compute_on_beam():
     runner = fn_api_runner.FnApiRunner()  # local runner
+    public_partitions=[int(partition) for partition in FLAGS.public_partitions] if FLAGS.public_partitions is not None else None
```
There is no need to split it, because we are using DEFINE_list, that returns list from a comma separated list. ([ref.](https://abseil.io/docs/python/guides/flags#flag-types))

dvadym replied:
Ah, cool!

lovroprepolec commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
+        class PartitionsFilterJoin(beam.DoFn):
+          def process(self, joined_data):
+            key, rest = joined_data
+            values, is_public = rest.get('values'), rest.get('is_public')
```
Done.

lovroprepolec commented:
Thanks for feedback Vadym. I made the changes you've suggested (except for the command line flags where it was not needed - see the response to your comment).
Let me know if there is anything else that should be improved. Thanks!

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,52 @@ def group_by_key(self, col, stage_name: str):
     def filter(self, col, fn, stage_name: str):
         return col | stage_name >> beam.Filter(fn)
 
+    def filter_by_key(self, col, public_partitions, data_extractors, stage_name: str):
```
Could you please implement filter_by_key in LocalOperations?

It will be very similar to BeamOperations, but no need to consider join case.

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/dp_engine.py:
```diff
@@ -48,7 +48,17 @@ def aggregate(self, col, params: AggregateParams,
         self._report_generators.append(ReportGenerator(params))
         # TODO: implement aggregate().
         # It returns input for now, just to ensure that the an example works.
```
Please drop this comment

lovroprepolec replied:
Done.

dvadym reviewed (changes requested):
It looks good. I left some comments.

Could you please run `make format` ? (it looks that some lines are longer that according to styleguide)

lovroprepolec commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,52 @@ def group_by_key(self, col, stage_name: str):
         return col | stage_name >> beam.Filter(fn)
 
+    def filter_by_key(self, col, public_partitions, data_extractors, stage_name: str):
+        """Filter out the partitions that are not not meant to be public.
```
Done.

lovroprepolec commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,52 @@ def group_by_key(self, col, stage_name: str):
+        col = col | "Mapping data by partition" >> beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        if isinstance(public_partitions, (list, set)):
+            public_partitions = set(public_partitions) if not isinstance(public_partitions, set) else public_partitions
```
Done.

lovroprepolec commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +90,52 @@ def group_by_key(self, col, stage_name: str):
+                    for value in values:
+                        yield key, value
+
+        def is_public(col):
```
Done.

lovroprepolec commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,63 @@
+    def test_filter_by_key_pcollection_noop(self):
+        col = [(1, 6, 1), (2, 7, 1), (3, 6, 1), (4, 7, 1), (5, 8, 1)]
+        public_partitions = []
+        with beam.Pipeline() as p:
```
Thanks. Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -38,6 +38,21 @@ def group_by_key(self, col, stage_name: str):
+        """Filter out the partitions that are not not meant to be public.
+
+        Args:
+          col: input collection
```
collection with elements (partition_key, data).

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -38,6 +38,21 @@ def group_by_key(self, col, stage_name: str):
 
+    @abc.abstractmethod
+    def filter_by_key(self, col, public_partitions, stage_name: str):
+        """Filter out the partitions that are not not meant to be public.
```
Filter -> Filters
and maybe simpler:
"Filters out not public partitions."

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -38,6 +38,21 @@ def group_by_key(self, col, stage_name: str):
+
+        Args:
+          col: input collection
+          public_partitions: collection of public partition keys
```
Please add '.' in the end of the comments

lovroprepolec replied:
Done.

dvadym commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,64 @@
+        col = [(1, 6, 1), (2, 7, 1), (3, 6, 1), (4, 7, 1), (5, 8, 1)]
+        public_partitions = []
+        result = self.ops.filter_by_key(col, public_partitions, self.data_extractors, "Public partition filtering")
+        assert result == []
```
Please replace assert here and below with
self.assertEqual (or other self.assert<something>). More details [here](https://docs.python.org/3/library/unittest.html) and some examples in this file below

lovroprepolec replied:
Thx. Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +101,42 @@ def group_by_key(self, col, stage_name: str):
+            if not isinstance(public_partitions, set):
+                public_partitions = set(public_partitions)
+            return col | "Filtering data from public partitions" >> beam.Filter(has_public_partition_key)
+        else:
```
Nit: drop else (it will be less nesting)

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +101,42 @@ def group_by_key(self, col, stage_name: str):
+        col = col | "Mapping data by partition" >> beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        if isinstance(public_partitions, (list, set)):
+            if not isinstance(public_partitions, set):
```
Please add comment before this line:
\# Public partitions are in memory.

lovroprepolec replied:
Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -86,6 +101,42 @@ def group_by_key(self, col, stage_name: str):
+                public_partitions = set(public_partitions)
+            return col | "Filtering data from public partitions" >> beam.Filter(has_public_partition_key)
+        else:
+            public_partitions = public_partitions | "Creating public_partitions PCollection" >> beam.Map(lambda x: (x, True))
```
Please add comment
\# Public partitions are not in memory. Filter out with a join.

lovroprepolec replied:
Done.

dvadym reviewed (commented):
Some minor comments.

It looks good, it's left only `filter_by_key` for LocalOperations.

lovroprepolec commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,64 @@
+        col = [(1, 6, 1), (2, 7, 1), (3, 6, 1), (4, 7, 1), (5, 8, 1)]
+        public_partitions = []
+        with test_pipeline.TestPipeline() as p:
+            pcol = (p | beam.Create(col))
```
() is needed for the test to run. Otherwise an AttributeError is raised:
```
    with test_pipeline.TestPipeline as p:
AttributeError: __enter__
```

dvadym replied:
Ack

lovroprepolec commented:
Hey Vadym,

I've implemented the LocalOperations filter_by_key, and added tests for it. I've also run the `make format`.

Please have a look, and let me know if anything else is needed. Thanks!

Lovro

dvadym commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,76 @@
+            privacy_id_extractor=lambda x: x[0],
+            value_extractor=lambda x: x[2])
+
+    def test_filter_by_key_must_not_be_none(self):
```
tests in-memory vs join-based are almost duplicated. It's good to avoid duplication especially if it's not hard. A good thing that there is a good solution - test parametrization, namely each test might have 1 bool parameter, say in_memory, if true then to run in-memory and false join-based

We can use absl parametrization, there are a few exampels [here](https://github.com/abseil/abseil-py/blob/master/absl/testing/parameterized.py).

Could you please make them parametrizable with absl? LMK if you have any questions.

lovroprepolec replied:
Thanks for suggesting that. Done.

dvadym commented on pipeline_dp/pipeline_operations.py:
```diff
@@ -131,6 +193,13 @@ def group_by_key(self, rdd, stage_name: str = None):
+                      public_partitions,
+                      data_extractors,
+                      stage_name: str = None):
+        pass
```
Nit: raise NotImplementedError("filter_by_key is not implemented in SparkRDDOperations")

lovroprepolec replied:
Done.

dvadym reviewed (commented):
Thanks, a few minor comments!

lovroprepolec commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,76 @@
+                                            self.data_extractors,
+                                            "Public partition filtering")
+
+    def test_filter_by_key_noop(self):
```
Sure, thanks. This was a leftover from the change of logic we did (empty filter -> empty output was earlier empty filter -> full output).

lovroprepolec replied:
Done.

dvadym commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,83 @@
+        {'in_memory': False},
+    )
+    def test_filter_by_key_remove(self, in_memory):
+        col = [(1, 7, 1), (2, 19, 1), (3, 9, 1), (4, 11, 1), (5, 10, 1)]
```
For BeamOperations, `col` should be PCollection as well. i.e. please apply beam.Create() here and in other places

lovroprepolec replied:
Done, for all tests. Looks much better now, thanks!

dvadym commented on tests/pipeline_operations_test.py:
```diff
@@ -1,14 +1,83 @@
+            self.assertEqual(result, expected_result)
+        else:
+            with test_pipeline.TestPipeline() as p:
+                pcol = (p | beam.Create(col))
```
in_memory is only about public_partitions, please replace pcol -> public_partitions.

And then parametrized tests will be simplified, this line will be the only difference in tests.

lovroprepolec replied:
Got it. Thanks. Done.
